chore: add 'remove liquidity' fees

0xApotheosis opened the pull request:
## Description

- Adds gas fees for remove liquidity estimates
- Uses the `AssetInput` component to display current LP asset balance

What this does not do:

- Handle clicking "Remove liquidity", and any subsequent flows

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Contributes to https://github.com/shapeshift/web/issues/5960

## Risk

Small/medium - mainly UI related work, though has some account-related logic (primarily the derivation of account from quote params + lp pool data).

> What protocols, transaction types or contract interactions might be affected by this PR?

Related to THORChain LP withdrawal transactions, though does not execute them yet.

## Testing

For an existing position, enter an amount to remove and ensure that:

a) the gas fee value calculates correctly
b) the calculated fee is similar (the same as?) to that calculated in "Add Liquidity".

### Engineering

☝️

### Operations

☝️

## Screenshots (if applicable)

<img width="368" alt="Screenshot 2024-02-12 at 12 23 30 pm" src="https://github.com/shapeshift/web/assets/97164662/986c800f-6133-4f26-bd2d-aa47e9568ca0">

0xApotheosis commented:
Current dependencies on/for this PR:
* `develop`
  * **PR #6175** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6175?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/6175?utm_source=stack-comment).

woodenfurniture commented on src/pages/ThorChainLP/Position/Position.tsx:
```diff
@@ -407,10 +407,13 @@ export const Position = () => {
+                      opportunityId={foundPool.opportunityId}
+                      poolAccountId={params.poolAccountId}
+                    />
+                  )}
```
is there an issue having an empty tab panel?

0xApotheosis replied:
You're right, not ideal. Fixed in https://github.com/shapeshift/web/pull/6175/commits/55d61881b96485d99aa59c851bf113e41c291381.

woodenfurniture commented on src/pages/ThorChainLP/components/RemoveLiquidity/RemoveLiquidityInput.tsx:
```diff
@@ -391,22 +581,26 @@ export const RemoveLiquidityInput: React.FC<RemoveLiquidityProps> = ({
+              percentOptions={percentOptions}
+              isReadOnly={false}
+              formControlProps={formControlProps}
+            ></AssetInput>
```
```suggestion
            />
```

0xApotheosis replied:
https://github.com/shapeshift/web/pull/6175/commits/55d61881b96485d99aa59c851bf113e41c291381

woodenfurniture commented on src/pages/ThorChainLP/components/RemoveLiquidity/RemoveLiquidityInput.tsx:
```diff
@@ -186,6 +364,28 @@ export const RemoveLiquidityInput: React.FC<RemoveLiquidityProps> = ({
+    const transactionType = getThorchainLpTransactionType(poolAsset.chainId)
+
+    switch (transactionType) {
+      case 'MsgDeposit': {
+        return THORCHAIN_POOL_MODULE_ADDRESS
+      }
+      case 'EvmCustomTx': {
+        // TODO: this should really be inboundAddressData?.router, but useQuoteEstimatedFeesQuery doesn't yet handle contract calls
+        // for the purpose of naively assuming a send, using the inbound address instead of the router is fine
+        return inboundAddressesData?.address
+      }
+      case 'Send': {
+        return inboundAddressesData?.address
+      }
+      default: {
+        assertUnreachable(transactionType as never)
+      }
```
the braces around switch cases are normally to allow defining new variables within a case, so shouldn;'t be required here

0xApotheosis replied:
That's fair - fixed in https://github.com/shapeshift/web/pull/6175/commits/55d61881b96485d99aa59c851bf113e41c291381.

woodenfurniture commented on src/pages/ThorChainLP/components/RemoveLiquidity/RemoveLiquidityInput.tsx:
```diff
@@ -154,8 +276,64 @@ export const RemoveLiquidityInput: React.FC<RemoveLiquidityProps> = ({
+
+  const { data: estimatedPoolAssetFeesData, isLoading: isEstimatedPoolAssetFeesDataLoading } =
+    useQuoteEstimatedFeesQuery({
+      collateralAssetId: poolAsset?.assetId ?? '', // fixme
```
are these fixmes still needed?

0xApotheosis replied:
They are not - improvement comments in https://github.com/shapeshift/web/pull/6175/commits/f8bb0b3f9803b822b907745861b5cb318940a4ab to explain why they might be `undefined`.

woodenfurniture commented:
# Testing notes

✅ Gas estimate appearing in UI
✅ Gas estimate refetches when changing remove amount %
✅ Gas estimate refetches when changing remove amount crypto value
✅ Gas estimate refetches when changing remove amount fiat value
✅ Fiat/Crypto toglging working as expected
✅ Gas estimate cached

---

Network hammering while dragging slider - needs debounce

https://github.com/shapeshift/web/assets/125113430/dd38343b-bded-4937-a0a6-06038eb0ff42

---

Loading skeleton missing after first load of gas, when updating the gas estimate

https://github.com/shapeshift/web/assets/125113430/8279dadd-c066-4b69-8d43-9fa69608e112

---

Gas estimate doest seem to poll or update at all

![image](https://github.com/shapeshift/web/assets/125113430/1a2c6e9e-bf37-4863-a3f7-57ba582dd8b7)

---

Hover state on the card for some reason

https://github.com/shapeshift/web/assets/125113430/36e7f7f2-fab6-4f04-8841-4e39acf094b0

---

## Unsure / Non-blockers / Unrelated to this PR

Clicking "remove liquidity" shows placeholder text

https://github.com/shapeshift/web/assets/125113430/72279265-cd94-4a21-b9d9-825bc586cc0e

---

No loading skeleton for gas estimate during loading. Gas estimate goes to 0 and returns after loading

https://github.com/shapeshift/web/assets/125113430/c3f6cbca-6c02-48b8-9e6a-d0314bc748ef

woodenfurniture reviewed (approved):
LGTM noting many of the larger feedback points will be addressed in a follow up PR

gomesalexandre commented on src/pages/ThorChainLP/components/RemoveLiquidity/RemoveLiquidityInput.tsx:
```diff
@@ -99,16 +218,19 @@ export const RemoveLiquidityInput: React.FC<RemoveLiquidityProps> = ({
     )
     if (!_UserlpData) return
     setUserlpData(_UserlpData)
+    const runeAddress = _UserlpData?.runeAddress
```
Wrong casing in `_UserlpData` on l. 216?

0xApotheosis replied:
Nice catch: https://github.com/shapeshift/web/pull/6175/commits/c13578dbbaf318dd43792431b655e7f7601d26a5

gomesalexandre commented on src/pages/ThorChainLP/components/RemoveLiquidity/RemoveLiquidityInput.tsx:
```diff
@@ -87,10 +106,110 @@ export const RemoveLiquidityInput: React.FC<RemoveLiquidityProps> = ({
+  const [virtualRuneFiatLiquidityAmount, setVirtualRuneFiatLiquidityAmount] = useState<
+    string | undefined
+  >()
+  const [cryptoAssetBalance, setCrytoAssetBalance] = useState<string | undefined>()
```
And `setCrytoRuneBalance` too
```suggestion
  const [cryptoAssetBalance, setCryptoAssetBalance] = useState<string | undefined>()
```

gomesalexandre replied:
Also we probably want to use a different terminology here? i.e balance would refer to the wallet balance, perhaps `stakedAmountCryptoBaseUnit` ?

Or even better, remove these and don't set anything in state since this is simply destructuring and renaming from userLpData and setting properties such as `userLpData.underlyingAssetAmountCryptoPrecision` as state values?

0xApotheosis replied:
gm https://github.com/shapeshift/web/pull/6175/commits/c13578dbbaf318dd43792431b655e7f7601d26a5

gomesalexandre replied:
Do we want to remove these altogether here?

0xApotheosis replied:
That's reasonable 🙏  https://github.com/shapeshift/web/pull/6175/commits/ce65270269aca06920532f2bb741e003db7495dd

gomesalexandre commented on src/pages/ThorChainLP/components/RemoveLiquidity/RemoveLiquidityInput.tsx:
```diff
@@ -154,8 +276,64 @@ export const RemoveLiquidityInput: React.FC<RemoveLiquidityProps> = ({
         .times(isAsymAssetSide || isAsymRuneSide ? 2 : 1)
         .toFixed(),
     )
+    setCrytoAssetBalance(underlyingAssetAmountCryptoPrecision)
+    setFiatAssetBalance(underlyingAssetValueFiatUserCurrency)
+    setCrytoRuneBalance(underlyingRuneAmountCryptoPrecision)
+    setFiatRuneBalance(underlyingRuneValueFiatUserCurrency)
```
As mentioned above, these should be removed as they're simply re-setting properties from `userLpData`

0xApotheosis replied:
https://github.com/shapeshift/web/pull/6175/commits/ce65270269aca06920532f2bb741e003db7495dd

gomesalexandre commented on src/pages/ThorChainLP/components/RemoveLiquidity/RemoveLiquidityInput.tsx:
```diff
@@ -87,28 +106,131 @@ export const RemoveLiquidityInput: React.FC<RemoveLiquidityProps> = ({
   const [userlpData, setUserlpData] = useState<UserLpDataPosition | undefined>()
+  const [runeAccountId, setRuneAccountId] = useState<AccountId | undefined>()
   const [percentageSelection, setPercentageSelection] = useState<number>(50)
+  const [shareOfPoolDecimalPercent, setShareOfPoolDecimalPercent] = useState<string | undefined>()
```
Not using this anywhere for remove liquidity

0xApotheosis replied:
<img width="346" alt="Screenshot 2024-02-21 at 9 15 57 am" src="https://github.com/shapeshift/web/assets/97164662/ec080523-af4d-4465-bb1a-e31b1b98b3db">

We _do_ have it on the reusable LP confirm component (available in the PR above this). I'm not sure how useful it is, but it's there for now.

gomesalexandre reviewed (commented):
First pass, tested locally - LGTM overall, though noticed some issues on the performance side of things as well as wrong underlying balances for asym withdraws.

Also some comment still standing WRT state fields cleanup.

### Asym asset side

- Starting balance

<img width="1391" alt="image" src="https://github.com/shapeshift/web/assets/17035424/256451bb-e70c-41bf-bad1-a77b4c2b1cca">
<img width="985" alt="image" src="https://github.com/shapeshift/web/assets/17035424/35e76667-df5c-438d-9828-6c2b2314c4f3">
<img width="468" alt="image" src="https://github.com/shapeshift/web/assets/17035424/12e55097-59fd-4841-9294-62010e2dc970">

- 100 % removal 

<img width="422" alt="image" src="https://github.com/shapeshift/web/assets/17035424/eb1173e6-4a43-435d-8aed-59ddd10c708c">

Note the balance (`Balance 0.12193277 AVAX`) is half of what the 100% removal allows (`0.24386553727977818994 AVAX`), which doesn't make sense - since this is an asym pool, the AVAX balance should be multiplied by two, i.e not showing the asset portion of the LP only, but the full, non-rebalanced balance, prorated in asset

- XHRs/loading state

As mentioned by @woodenfurniture previously, we definitely want to add skeletons as well as debouncing here, as sliding over spams XHRs

- Estimated gas fees look sane

<img width="445" alt="image" src="https://github.com/shapeshift/web/assets/17035424/ed1dbbe9-edf7-45c2-8f01-e89d8226a6f3">

### Asym RUNE side

- Ditto here, balance should be prorated in RUNE, i.e not showing only the RUNE side of the rebalanced deposit, but the original RUNE deposit, see 100% withdraw which is twice the balance

<img width="465" alt="image" src="https://github.com/shapeshift/web/assets/17035424/cf69d904-37fb-412b-9d67-b701658a8cfd">
<img width="440" alt="image" src="https://github.com/shapeshift/web/assets/17035424/16ec4955-2fc6-41df-994e-267afec6657e">

- Fees look sane

<img width="388" alt="image" src="https://github.com/shapeshift/web/assets/17035424/3cda7b00-6be4-45eb-bc80-83c9802a1ff9">

- Slippage is visually borked and recomputes as requests are going through

https://github.com/shapeshift/web/assets/17035424/7a0c4ed5-821c-441e-a6d2-2c1b1820c69e

In addition to a skeleton which would alleviate the loading state side of things, it looks like `estimateRemoveThorchainLiquidityPosition` is fired too many times (i.e 4 calls to `/v2/pool/<pool>` for each slippage calculation.

### Sym

- Starting balance

<img width="307" alt="image" src="https://github.com/shapeshift/web/assets/17035424/90852757-c8ee-400d-ab28-dfb77d143e12">

- 100% withdraw looks sane

<img width="418" alt="image" src="https://github.com/shapeshift/web/assets/17035424/cdc3eaeb-d286-467c-be1a-f6fce85e4b49">

- Gas fees look sane

<img width="383" alt="image" src="https://github.com/shapeshift/web/assets/17035424/85c12c81-005b-41f3-b478-d142e126bc6a">

0xApotheosis commented:
Thanks for the stellar review @gomesalexandre!

Fixed the amount logic so it actually shows the correct amounts for sym and asym pools.
Also fixed your smaller comments.

I'll merge this and do a follow-up PR addressing the loading skeletons and network request feedback.
